File: Source/Seasonal/Doodads/DoodadChanger.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class DoodadChanger
{
    private static readonly int SafezoneLanterns = FourCC("B005");
    private static readonly List<destructable> AllDestructables = new();
    private static SeasonTheme _theme = SeasonThemeRegistry.None;

    public static void Initialize()
    {
        CreateInitDestructables();
        if (Gamemode.CurrentGameMode != GameMode.Standard) return;
        Apply(SeasonalManager.CurrentTheme);
    }

    /// <summary>
    /// Swaps the safezone decoration for the theme's version and shows/hides its
    /// ambient decor. Replaces the old ChristmasDoodads()/NoSeasonDoodads() pair -
    /// works for any theme, including future seasons.
    /// </summary>
    public static void Apply(SeasonTheme theme)
    {
        _theme = theme;
        ReplaceSafezoneDecor(theme.SafezoneDecorType, theme.SafezoneDecorScale);
        ShowSeasonalDoodads(theme.Season != HolidaySeasons.None);
    }

    // Compatibility wrappers - remove once you've confirmed nothing outside
    // these files calls them directly.
    public static void NoSeasonDoodads() => Apply(SeasonThemeRegistry.None);
    public static void ChristmasDoodads() => Apply(SeasonThemeRegistry.Christmas);

    private static void ReplaceSafezoneDecor(int newType, float scale)
    {
        List<(float x, float y)> positions = new();

        foreach (var des in AllDestructables)
        {
            positions.Add((des.X, des.Y));
            des.Dispose();
        }

        AllDestructables.Clear();

        foreach (var pos in positions)
        {
            var newDestructible = destructable.CreateDead(newType, pos.x, pos.y, 0, scale);
            AllDestructables.Add(newDestructible);
        }
        GC.RemoveList(ref positions);
    }

    private static void CreateInitDestructables()
    {
        var counter = 0;

        foreach (var safeZone in Globals.SAFE_ZONES)
        {
            var rect = safeZone.Rect_;

            var minX = rect.MinX;
            var minY = rect.MinY;
            var maxX = rect.MaxX;
            var maxY = rect.MaxY;

            if (counter % 4 != 0)
                AllDestructables.Add(destructable.CreateDead(SafezoneLanterns, minX, maxY)); // Top left corner
            if (counter % 4 != 1 && counter != 14)
                AllDestructables.Add(destructable.CreateDead(SafezoneLanterns, maxX, maxY)); // Top right corner
            if (counter % 4 != 2)
                AllDestructables.Add(destructable.CreateDead(SafezoneLanterns, maxX, minY)); // Bottom right corner
            if (counter % 4 != 3 && counter != 0)
                AllDestructables.Add(destructable.CreateDead(SafezoneLanterns, minX, minY)); // Bottom left corner

            counter++;
        }
    }

    public static void ShowSeasonalDoodads(bool show = false) =>
        EnumDestructablesInRect(Globals.WORLD_BOUNDS, null, () => HideDoodads(show));

    private static void HideDoodads(bool show)
    {
        var des = GetEnumDestructable();

        // Only touch destructables that are seasonal decor for SOME theme - leave
        // unrelated map doodads alone. Checking the full registry (not just the
        // active theme) is what makes this correct when switching AWAY from a
        // season: the active theme's own list is empty at that point, so checking
        // only against it would never find anything to hide.
        if (!SeasonThemeRegistry.IsSeasonalDecor(des.Type)) return;

        Console.WriteLine("seasonal decor: " + des.Type + " belongs to active theme: " + ContainsType(_theme.DecorTypes, des.Type));
        Console.WriteLine(show ? "showing" : "hiding");
        var belongsToActiveTheme = ContainsType(_theme.DecorTypes, des.Type);
        des.SetVisibility(show && belongsToActiveTheme);
    }

    private static bool ContainsType(int[] types, int type)
    {
        for (var i = 0; i < types.Length; i++)
        {
            if (types[i] == type) return true;
        }
        return false;
    }
}

File: src/Seasonal/Doodads/DoodadChanger.ts
import { CurrentGameMode } from 'src/Gamemodes/CurrentGameMode'
import { GameMode } from 'src/Gamemodes/GameModeEnum'
import { Globals } from 'src/Global/Globals'
import { GC } from 'src/Utility/GC'
import { Destructable } from 'w3ts'
import { HolidaySeasons, Seasons } from '../Seasons'

export class DoodadChanger {
    private static SafezoneLanterns = FourCC('B005')
    private static ChristmasTree = FourCC('B001')
    private static CrystalRed = FourCC('B002')
    private static CrystalBlue = FourCC('B003')
    private static CrystalGreen = FourCC('B004')
    private static Snowglobe = FourCC('B006')
    private static Lantern = FourCC('B007')
    private static Fireplace = FourCC('B008')
    private static Snowman = FourCC('B009')
    private static Firepit = FourCC('B00A')
    private static Igloo = FourCC('ITig')
    private static RedLavaCracks = FourCC('B00B')
    private static BlueLavaCracks = FourCC('B00C')
    private static SuperChristmasTree = FourCC('B00D')
    public static ChristmasDecor: number[]
    private static AllDestructables: destructable[] = []

    public static Initialize = () => {
        DoodadChanger.CreateInitDestructiables()
        if (CurrentGameMode.active !== GameMode.Standard) return
        DoodadChanger.SeasonalDoodads()
    }

    public static InitChristmasDecor = () => {
        return [
            DoodadChanger.CrystalRed,
            DoodadChanger.CrystalBlue,
            DoodadChanger.CrystalGreen,
            DoodadChanger.Snowglobe,
            DoodadChanger.Lantern,
            DoodadChanger.Fireplace,
            DoodadChanger.Snowman,
            DoodadChanger.Firepit,
            DoodadChanger.Igloo,
            DoodadChanger.RedLavaCracks,
            DoodadChanger.BlueLavaCracks,
            DoodadChanger.SuperChristmasTree,
        ]
    }

    private static SeasonalDoodads = () => {
        DoodadChanger.ChristmasDoodads()
    }

    public static NoSeasonDoodads = () => {
        DoodadChanger.ReplaceDoodad(DoodadChanger.SafezoneLanterns, 1.0)
        DoodadChanger.ShowSeasonalDoodads(false)
    }

    public static ChristmasDoodads = () => {
        if (Seasons.getCurrentSeason() !== HolidaySeasons.Christmas) return
        DoodadChanger.ReplaceDoodad(DoodadChanger.ChristmasTree, 2.5)
        DoodadChanger.ShowSeasonalDoodads(true)
    }

    private static ReplaceDoodad = (newType: number, scale: number) => {
        const positions: { x: number; y: number }[] = []

        for (const des of DoodadChanger.AllDestructables) {
            positions.push({ x: GetDestructableX(des), y: GetDestructableY(des) })
            RemoveDestructable(des)
        }

        DoodadChanger.AllDestructables = []
        for (const pos of positions) {
            const newDestructible = CreateDeadDestructable(newType, pos.x, pos.y, 0, scale, 0)!
            DoodadChanger.AllDestructables.push(newDestructible)
        }
        GC.RemoveList(positions)
    }

    private static CreateInitDestructiables = () => {
        let counter = 0

        for (const safeZone of Globals.SAFE_ZONES) {
            const rect = safeZone.Rectangle

            const minX = rect.minX
            const minY = rect.minY
            const maxX = rect.maxX
            const maxY = rect.maxY

            if (counter % 4 !== 0) {
                const des = CreateDeadDestructable(DoodadChanger.SafezoneLanterns, minX, maxY, 0, 1, 0)
                if (des) DoodadChanger.AllDestructables.push(des) // Top left corner
            }
            if (counter % 4 !== 1 && counter !== 14) {
                const des = CreateDeadDestructable(DoodadChanger.SafezoneLanterns, maxX, maxY, 0, 1, 0)
                if (des) DoodadChanger.AllDestructables.push(des) // Top right corner
            }
            if (counter % 4 !== 2) {
                const des = CreateDeadDestructable(DoodadChanger.SafezoneLanterns, maxX, minY, 0, 1, 0)
                if (des) DoodadChanger.AllDestructables.push(des) // Bottom right corner
            }
            if (counter % 4 !== 3 && counter !== 0) {
                const des = CreateDeadDestructable(DoodadChanger.SafezoneLanterns, minX, minY, 0, 1, 0)
                if (des) DoodadChanger.AllDestructables.push(des) // Bottom left corner
            }

            counter++
        }
    }

    public static ShowSeasonalDoodads = (show = false) => {
        return EnumDestructablesInRect(Globals.WORLD_BOUNDS.handle, null as never, () =>
            DoodadChanger.HideDoodads(show)
        )
    }

    private static HideDoodads = (show: boolean) => {
        const des = Destructable.fromHandle(GetEnumDestructable())!
        if (DoodadChanger.ChristmasDecor.includes(des.typeId)) des.show(show)
    }
}
